onnxruntime/core/optimizer/computation_reduction.cc

```diff
@@ -0,0 +1,228 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/computation_reduction.h"
+#include "core/graph/graph_utils.h"
+
+using namespace ONNX_NAMESPACE;
+using namespace ::onnxruntime::common;
+namespace onnxruntime {
+typedef std::function<Status(Graph&, Node&, Node&)> Handler;
+
+static const int GATHERND_BATCH_DIM = 1;
+
+static bool IsLeadingDimsEqual(const TensorShapeProto* input_shape, const TensorShapeProto* output_shape,
+                               int num_dim_to_check) {
+  if (output_shape->dim_size() < num_dim_to_check || input_shape->dim_size() < num_dim_to_check) {
+    return false;
+  }
+
+  for (int i = 0; i < num_dim_to_check; i++) {
+    auto& output_dim = output_shape->dim(i);
+    auto& input_dim = input_shape->dim(i);
+    if (output_dim.has_dim_value() && input_dim.has_dim_value()) {
+      if (output_dim.dim_value() != input_dim.dim_value()) {
+        return false;
+      }
+    } else if (output_dim.has_dim_param() && input_dim.has_dim_param()) {
+      if (output_dim.dim_param() != input_dim.dim_param()) {
+        return false;
+      }
+    } else {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+static int GetInputNodeToReplace(const Node& gathernd_input_node) {
+  // If gathernd_input_node's some input tensor have exactly same shape with
+  // gathernd_input_node output tensor shape, then it is safe to gather using
+  // original slice ranges.
+  int gathernd_input_index = -1;
+  auto output_shape = gathernd_input_node.OutputDefs()[0]->Shape();
+  const int output_rank = output_shape->dim_size();
+  for (size_t i = 0; i < gathernd_input_node.InputDefs().size(); i++) {
+    auto input_shape = gathernd_input_node.InputDefs()[i]->Shape();
+    const int input_rank = input_shape->dim_size();
+    if (input_rank != output_rank) {
+      continue;
+    }
+
+    // We compare the first GATHERND_BATCH_DIM + 1 to make sure this input node
+    // (of gathernd_input_node) can be input of GatherND.
+    if (IsLeadingDimsEqual(input_shape, output_shape, GATHERND_BATCH_DIM + 1)) {
+      gathernd_input_index = i;
```

Comment:
warning C4267: '=': conversion from 'size_t' to 'int', possible loss of data

Reply:
thanks :)